src/WebJobs.Extensions.CosmosDB/CosmosDBSqlResolutionPolicy.cs

```diff
@@ -36,18 +35,44 @@ public string TemplateBind(PropertyInfo propInfo, Attribute resolvedAttribute, B
             SqlParameterCollection paramCollection = new SqlParameterCollection();
 
             // also build up a dictionary replacing '{token}' with '@token' 
-            IDictionary<string, string> replacements = new Dictionary<string, string>();
+            IDictionary<string, object> replacements = new Dictionary<string, object>();
             foreach (var token in bindingTemplate.ParameterNames.Distinct())
             {
-                string sqlToken = $"@{token}";
-                paramCollection.Add(new SqlParameter(sqlToken, bindingData[token]));
-                replacements.Add(token, sqlToken);
+                object tokenObject = bindingData[token];
+                if (tokenObject is string tokenString)
+                {
+                    string sqlToken = GetSqlParameterName(token);
+                    paramCollection.Add(new SqlParameter(sqlToken, tokenString));
+                    replacements.Add(token, sqlToken);
+                }
+                else if (tokenObject is IDictionary<string, string> tokenDictionary)
+                {
+                    IDictionary<string, string> tokens = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+                    foreach (var item in tokenDictionary)
+                    {
+                        string fullToken = $"{token}.{item.Key}";
+                        bool tokenInUse = bindingTemplate.Pattern.Contains(fullToken);
+                        if (tokenInUse)
+                        {
+                            string sqlToken = GetSqlParameterName(fullToken);
+                            paramCollection.Add(new SqlParameter(sqlToken, item.Value));
+                            tokens.Add(item.Key, sqlToken);
+                        }
+                    }
+                    replacements.Add(token, tokens);
+                }
             }
 
-            docDbAttribute.SqlQueryParameters = paramCollection;
+            cosmosDBAttribute.SqlQueryParameters = paramCollection;
 
-            string replacement = bindingTemplate.Bind(new ReadOnlyDictionary<string, string>(replacements));
+            string replacement = bindingTemplate.Bind(new ReadOnlyDictionary<string, object>(replacements));
             return replacement;
         }
+
+        private string GetSqlParameterName(string name)
+        {
+            const string safeSeparator = "_";
+            return "@" + name.Replace(".", safeSeparator).Replace("-", safeSeparator);
```

Comment:
Are these not allowed in SqlParameter names? Is the list larger than this, or is it just these two characters? When doing replacements like this, I worry that we could stomp over another matching token. Even if it's rare, I wonder if there's another way to handle it.

Reply:
@ealsur do you know the complete naming rules for the parameters?

Good point. We can generate something like a GUID for the name to avoid clobbering an existing token. If we do that, as long as it fits under the length limit for a parameter name, it should work.

Reply:
@anthonychu  Sorry, I missed this for some reason, never saw the notification :( Regarding the rules, I'm not really familiar with all of them but [this one](https://stackoverflow.com/questions/751719/sql-server-invalid-characters-in-parameter-names) looks complete.

Reply:
Thanks. So I guess Cosmos DB follows SQL rules for param names? Perhaps switching to a GUID is a safer way to go. I'll see if I can do that.